### PR TITLE
fix(braces): F+ for ; after {} in struct/union reference

### LINKED ISSUE #201: [readability/braces] False positive for ; after {}
```c
    struct mount_attr *attr = &(struct mount_attr){};
```

I'm not sure if this is valid C++.  It is valid C, for sure.

```
tmp/src/man2/mount_setattr.2.d/mount_setattr.c:61:  You don't need a ; after a }  [readability/braces] [4]
```

The report in this case is invalid, since the `;` is necessary.

### PR DESCRIPTION
**Pull request by aaronliu0130**
Immunize lines with `&(struct` in them

Should fix #201

### CODE REVIEW

File: cpplint_unittest.py [hunk 1]
```diff
@@ -3062,6 +3062,7 @@ def testSemiColonAfterBraces(self):
                     for identifier in ["", " x"]:
                         self.TestLint(keyword + align + typename + " {}" + identifier + ";", "")
 
+        self.TestLint("struct mount_attr *attr = &(struct mount_attr){};", "")
```
[the remark this answers is not in the document]
Comment (aaronliu0130) [reply cpplint_unittest.py:3065]: Throughout the unit tests, each part of the relevant regex pattern is only tested once.

File: cpplint.py [hunk 2]
```diff
@@ -5055,6 +5055,7 @@ def CheckTrailingSemicolon(filename, clean_lines, linenum, error):
         #  - Compound literals
         #  - Lambdas
         #  - alignas specifier with anonymous structs
+        #  - references to anonymous struct
```
Comment (aaronliu0130) [inline cpplint.py]:
```suggestion
        #  - references to anonymous struct or union
```

**Review by janmarsino98 — CHANGES_REQUESTED** [review]
I reproduced the original issue and confirmed that the single-line `struct` case is fixed. The existing targeted test passes, the full functional suite passes (225 tests without the pre-existing coverage gate), and pylint, mypy, pre-commit, and the remote CI checks pass.

Requesting changes for the following issues:

1. `cpplint.py:5087` only recognizes `struct`/`union` when the keyword is on the same physical line as the opening parenthesis. C permits whitespace, including newlines, in the compound-literal type name. On this head, the following valid code still produces `You don't need a ; after a } [readability/braces] [4]`:

   ```c
   struct S *s = &(
       struct S
   ){};
   ```

   The same false positive remains for `union U`. Please make the detection work across the expression returned by `ReverseCloseExpression` and add multiline regression coverage.

2. `cpplint_unittest.py:3065` exercises only `struct`, although the implementation explicitly handles both `struct` and `union`. Please add a `union` compound-literal reference test and negative boundary cases such as `&(structure){}` and `&(struct_name){}`.

3. `CONTRIBUTING.rst` requires every PR to add a summary to `CHANGELOG.rst`, but this PR does not modify the changelog.

For context, running `pytest` with the repository's coverage configuration executes all 225 tests successfully but exits on the existing 89.36% coverage result versus the 90% threshold. I reproduced the identical result on base SHA `6734a6d`, so I am not treating that as a regression from this PR.

**Review by janmarsino98 — CHANGES_REQUESTED** [review]
Thanks for the re-review request! I took another look, but the PR is still on the same commit (a44637e) as my previous review, so the points I raised there are still outstanding: multiline struct/union handling and regression tests, the additional union/boundary cases, and the CHANGELOG.rst entry.

Once those updates are pushed, feel free to re-request my review and I'll be happy to take another look.

**Review by lovewave02 — CHANGES_REQUESTED** [review]
One valid C case is still uncovered. On this head, both of these still emit `You don't need a ; after a }`:

```c
const struct S *s = &(const struct S){};
volatile union U *u = &(volatile union U){};
```

`cc -std=c11 -fsyntax-only` accepts both. The new matcher only accepts a type name that begins directly with `struct` or `union`. Please handle type qualifiers in the compound-literal type name and add `const`/`volatile` regression cases.

**Comment by androvonx95** [comment]
I checked out a44637e and reproduced independently:

- Fixed: &(struct mount_attr){}; and &(union U){};
- Still FP: multiline &( struct S ){}; , &(const struct S){}; , &(volatile union U){};

Agree with the earlier review points: qualify detection across ReverseCloseExpression (and allow const/volatile), add union + qualifier + multiline tests, and add a CHANGELOG.rst TBA entry per CONTRIBUTING.rst.

**Review by yangfan-yf-yf — CHANGES_REQUESTED** [review]
There is another single-line compound-literal case that the new guard does not cover. The required semicolon is not specific to `struct` or `union` specifiers: C compound literals can use scalar and typedef type names as well. On `a44637e`, both of these still emit `You don't need a ; after a }`:

```c
int *value = &(int){0};

typedef struct S { int x; } S;
S *value = &(S){0};
```

Both semicolons are required. The typedef form has the same address-of compound-literal shape as the reported case; only the spelling of the type name differs. Please make the exemption recognize the address-of compound-literal form rather than only a `struct`/`union` prefix, and add scalar and typedef regression cases. The full test suite passes (`225 passed`), so these cases need focused coverage.

**Review by PNHD — CHANGES_REQUESTED** [review]
Re-reviewed current head a44637e. I independently confirmed the existing current-head blockers remain: valid compound literals outside the narrow single-line struct/union form are still mishandled, and the required CHANGELOG entry is still missing. I have no new inline findings to add beyond the existing reviews.

**Review by walidbi200 — CHANGES_REQUESTED** [review]
Reviewed current head `a44637e`. I reproduced the original #201 false positive and confirmed the current change fixes the narrow single-line `&(struct ...){};` / `&(union ...){};` forms, but the implementation is still tied to the spelling of the type name rather than the compound-literal structure.

Valid compound literals outside that narrow form still emit `readability/braces`, including:

    struct S *s = &(
        struct S
    ){};

    const struct S *s = &(const struct S){};
    volatile union U *u = &(volatile union U){};
    int *value = &(int){0};

and typedef-name forms such as `&(S_t){0}`.

I tested a more structural approach locally. Simply exempting every case where the text before the matching parenthesis ends in `&` is too broad: it suppresses the legitimate redundant-semicolon warning for function definitions such as `operator&`, `operator&&`, and conversion-to-reference operators.

A narrower version that recognizes the address-of compound-literal shape while excluding operator declarations handled the compound-literal matrix above without suppressing those operator-definition warnings in my tests.

The current PR is also still missing the CHANGELOG.rst entry required by CONTRIBUTING.rst.

Focused `SemiColonAfterBraces` regression tests, pre-commit, and `git diff --check` pass with the local candidate. Repository-wide pytest could not be completed in this environment because `testfixtures` is unavailable; the existing direct unit-file run also has an unrelated pre-existing quiet-mode warning failure.

I think the exemption should be generalized to the structural compound-literal case, with regression coverage for multiline, qualified, scalar, typedef, and operator controls, before this merges.
